extensions/pkg/controller/worker/genericactuator/actuator.go

```diff
@@ -281,6 +286,48 @@ func (a *genericActuator) shallowDeleteAllObjects(ctx context.Context, namespace
 	})
 }
 
+// IsMachineControllerStuck determines if the machine controller pod is stuck.
+// A pod is assumed to be stuck if
+//  - a machine deployment exists that does not have a machine set with the correct machine class
+//  - the machine set does not have a status that indicates (attempted) machine creation
+func (a *genericActuator) IsMachineControllerStuck(ctx context.Context, worker *extensionsv1alpha1.Worker) (bool, *string, error) {
+	machineDeployments := &machinev1alpha1.MachineDeploymentList{}
+	if err := a.client.List(ctx, machineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+		return false, nil, err
+	}
+
+	machineSets := &machinev1alpha1.MachineSetList{}
+	if err := a.client.List(ctx, machineSets, client.InNamespace(worker.Namespace)); err != nil {
+		return false, nil, err
+	}
+
+	isStuck, msg := isMachineControllerStuck(machineSets.Items, machineDeployments.Items)
+	return isStuck, msg, nil
+}
+
+func isMachineControllerStuck(machineSets []machinev1alpha1.MachineSet, machineDeployments []machinev1alpha1.MachineDeployment) (bool, *string) {
```

Comment:
Probably it makes sense to have some grace period of ~1-2m or so compared to the machine deployments/sets creation timestamps.

Reply:
It is only called after it has already unsuccessfully waited (3/5 minutes) for the machinedeployments to become ready. After this I think there is no need for a grace period any more?

Reply:
Hm yes, but wouldn't it be more convenient for the caller of this function to not care about these "grace periods" on his own? You are anyways reading the machine deployments/machine sets, so you can add the grace period behaviour this function so that callers aren't bothered with it.

Reply:
Mhm ok, to me that sounds like additional complexity as this check currently never fails. But I get where you coming from.

Maybe someone else has another opinion here. E.g. @prashanth26

Reply:
Why additional complexity? You can simply exit early/not consider machine deployments/sets that are too young. IMO, it's now more complex to demand from every caller to consider this grace period on its own.

Reply:
> If we do plan to have this logic, I would agree with @rfranzke on having the timeouts/grace-period inside the function for the caller to not care about it. 

Alright, will do.

> MachineControllerStuck is this something that was observed earlier? If yes, how often was it? Because if we have something like this, we have a serious problem at the MCM and need to fix it.

Observed couple of times (I also have seen this personally sometimes), thats why I tackled it as part of my quality issues.
But I agree, that should be fixed in the MCM. It is just very difficult to reproduce due to its sporadic nature.

> with support for scale to zero, the worker-pool might be created and the replicas could still be 0 right?

Makes sense. I will remove the check for MachineSets having a status.
So the only check would be if the necessary MachineSet with the right MachineClass got created.
Any other ideas?

>  check on the liveliness probe on the MCM

The MCM pod is already deployed with a liveness probe that is performed by the kubelet. Is that what you had in mind? 

Or are you suggesting that there should be a probe build into the provider extension? 
I am not sure if that is the right place for it.

Reply:
> Observed couple of times (I also have seen this personally sometimes), thats why I tackled it as part of my quality issues.

But I agree, that should be fixed in the MCM. It is just very difficult to reproduce due to its sporadic nature.
Okay. This is scaring me now. Please ping me if and when you see this ever again.

> Makes sense. I will remove the check for MachineSets having a status.
So the only check would be if the necessary MachineSet with the right MachineClass got created.
Any other ideas?

Yes, this should be enough for now. We can improve the liveliness checks on MCM if we see this reoccurring. 

> The MCM pod is already deployed with a liveness probe that is performed by the kubelet. Is that what you had in mind?

Yeah the ones probed by kubelet. But maybe improve the tasks to be performed by MCM during a liveliness check.

Reply:
> Okay. This is scaring me now. Please ping me if and when you see this ever again.

Sure will do. 

> We can improve the liveliness checks on MCM if we see this reoccurring. 

ok makes sense.

Reply:
These changes look good to me.  Thanks.